test/goerli-exchange-tests.js

```diff
@@ -0,0 +1,393 @@
+          0,
+        )
+      await expect(tx).to.revertedWith('Exchange / Unauthorized Caller')
+    })
+
+    describe('when transferring an exact amount to the exchange', async function () {
+      let localSnapshotId, initialWethWalletBalance
+
+      this.beforeEach(async function () {
+        localSnapshotId = await provider.send('evm_snapshot', [])
+
+        let tx = await WETH.deposit({
+          value: amountToWei(amount.toNumber()).toFixed(0),
+        })
+
+        const intBal = await balanceOf( MAINNET_ADRESSES.ETH,address);
+
+        initialWethWalletBalance = convertToBigNumber(
+          intBal
+        )
+
+        await WETH.approve(exchange.address, amountInWei)
+
+        await exchange.swapTokenForDai(
+          MAINNET_ADRESSES.ETH,
+          amountInWei,
+          receiveAtLeastInWei,
+          to,
+          data,
+          {
+            value: 0,
+            gasLimit: 2500000,
+          },
+        )
+
+      })
+
+      this.afterEach(async function () {
+        await provider.send('evm_revert', [localSnapshotId])
+      })
+
+      it(`should receive at least amount specified in receiveAtLeast`, async function () {
+        const wethBalance = convertToBigNumber(await balanceOf(MAINNET_ADRESSES.ETH, address))
+        const daiBalance = convertToBigNumber(await balanceOf(MAINNET_ADRESSES.MCD_DAI, address))
+
+        expect(wethBalance.toString()).to.equals(
+          initialWethWalletBalance.minus(amountToWei(amount)).toString(),
+        )
+        expect(daiBalance.gte(convertToBigNumber(receiveAtLeastInWei))).to.be.true
+      })
+
+      it('should have collected fee', async function () {
+        const beneficiaryDaiBalance = convertToBigNumber(
+          await balanceOf(MAINNET_ADRESSES.MCD_DAI, feeBeneficiary),
+        )
+
+        const expectedCollectedFee = amountFromWei(receiveAtLeastInWei)
+          .div(ONE.minus(slippage.asDecimal))
+          .times(fee.asDecimal)
+
+        expect(amountFromWei(beneficiaryDaiBalance).toFixed(6)).to.be.eq(
+          expectedCollectedFee.toFixed(6),
+        )
+      })
+    })
+
+    describe('when transferring less amount to the exchange', async function () {
+      let initialWethWalletBalance, lessThanTheTransferAmount, localSnapshotId
+
+      this.beforeEach(async function () {
+        localSnapshotId = await provider.send('evm_snapshot', [])
+
+        await WETH.deposit({
+          value: amountToWei(1000).toFixed(0),
+        })
+
+        initialWethWalletBalance = convertToBigNumber(
+          await balanceOf(MAINNET_ADRESSES.ETH, address),
+        )
+        lessThanTheTransferAmount = convertToBigNumber(amountInWei).minus(5);
+
+        await WETH.approve(exchange.address, lessThanTheTransferAmount.toFixed(0))
+      })
+
+      this.afterEach(async function () {
+        await provider.send('evm_revert', [localSnapshotId])
+      })
+
+      it('should throw an error and not exchange anything', async function () {
```

Comment:
This error will be thrown because o value 0 right ?

Reply:
actually because of approval of lessThanTheTransferAmount in beforeEach